common/thrift_router.h

```diff
@@ -368,69 +368,70 @@ class ThriftRouter {
+    auto selectHosts(
         const std::vector<std::pair<const Host*, Role>>& host_info,
         const Role role,
         const unsigned rotation_counter,
         const std::string& segment,
         const int shrink_target,
         const std::string& specific_az) {
-      std::vector<const Host*> v;
-      if (role == Role::ANY && !FLAGS_always_prefer_local_host) {
-        // prefer MASTER, then prefer groups with longer common prefix length
-        std::vector<const Host*> v_s;
-        for (const auto& hi : host_info) {
-          if (!matchAZ(hi.first->az, specific_az)) {
-            continue;
-          }
-          if (hi.second == Role::SLAVE) {
-            v_s.push_back(hi.first);
-          } else if (hi.second == Role::MASTER) {
+          std::vector<const Host*> v;
+          std::unordered_map<const Host*, Role> hostToRole;
+          v.reserve(host_info.size());
+          hostToRole.reserve(host_info.size());
+          for (const auto& hi : host_info) {
+            // filter by AZ
+            if (specific_az != "" && hi.first->az.compare(specific_az) != 0) {
+              continue;
+            }
+            // filter by role
+            if (role != Role::ANY && hi.second != role){
+              continue;
+            }
             v.push_back(hi.first);
+            hostToRole[hi.first] = hi.second;
           }
-        }
-        RankHostsByGroupPrefixLengthAndShrinkTo(&v, rotation_counter, segment, shrink_target);
-        RankHostsByGroupPrefixLengthAndShrinkTo(&v_s, rotation_counter, segment, shrink_target);
-        v.insert(v.end(), v_s.begin(), v_s.end());
-        if (shrink_target < v.size()) {
-          v.resize(shrink_target);
-        }
-      } else {
-        // prefer local
-        v.reserve(host_info.size());
-        for (const auto& hi : host_info) {
-          if (!matchAZ(hi.first->az, specific_az)) {
-            continue;
-          }
-          if (role == Role::ANY || hi.second == role) {
-            v.push_back(hi.first);
+          if (v.empty()) {
+            return v;
           }
-        }
-        RankHostsByGroupPrefixLengthAndShrinkTo(&v, rotation_counter, segment, shrink_target);
-      }
-      return v;
+
+          sortAndShrinkHosts(&v, hostToRole, rotation_counter, role, segment, shrink_target);
+          return v;
     }
 
-    void RankHostsByGroupPrefixLengthAndShrinkTo(
-        std::vector<const Host*>* v,
-        const unsigned rotation_counter,
-        const std::string& segment,
-        const int shrink_target) {
-      auto comparator = [this, &segment, rotation_counter]
+    void sortAndShrinkHosts(
+      std::vector<const Host*>* v,
+      const std::unordered_map<const Host*, Role>& hostToRole,
+      const unsigned rotation_counter,
+      const Role role,
+      const std::string& segment,
+      const int shrink_target) {
+        auto comparator = [this, hostToRole, role, &segment, rotation_counter]
```

Comment:
Can we pass `hostToRole` by reference? Will avoid a copy.

Reply:
done